std/variant.d

```diff
@@ -249,11 +249,21 @@ private:
         {
             static if (is(typeof(*rhsPA == *zis)))
             {
-                if (*rhsPA == *zis)
+                enum isEmptyStructWithoutOpEquals = (is(A == struct) && A.tupleof.length == 0 &&
+                                                     !__traits(hasMember, A, "opEquals"));
+                static if (!isEmptyStructWithoutOpEquals)
+                {
+                    if (*rhsPA == *zis)
+                        return 0;
+                }
+                static if (isEmptyStructWithoutOpEquals)
```

Comment:
```suggestion
                else
```

Reply:
This doen't work because there's an `else static if` after this `static if`.

Reply:
Not a big deal, but just seeing a condition followed by its negation seemed odd. The refactoring that would take care of that would be:

```d
                static if (isEmptyStructWithoutOpEquals)
                {
                    // The check above will always succeed if A is an empty struct.
                    // Don't generate unreachable code as seen in
                    // https://issues.dlang.org/show_bug.cgi?id=21231
                    return 0;
                }
                else
                {
                    if (*rhsPA == *zis)
                        return 0;
                    static if (is(typeof(*zis < *rhsPA)))
                    {
                        // Many types (such as any using the default Object opCmp)
                        // will throw on an invalid opCmp, so do it only
                        // if the caller requests it.
                        if (selector == OpID.compare)
                            return *zis < *rhsPA ? -1 : 1;
                        else
                            return ptrdiff_t.min;
                    }
                    else
                    {
                        // Not equal, and type does not support ordering
                        // comparisons.
                        return ptrdiff_t.min;
                    }
            }
```

This better clarifies the flow depending on compile-time conditions.

Reply:
done.